### DISCUSSION TITLE
Generate and Use Coverage at each run

### DISCUSSION THREAD

**Discussion by nordlow** [CyberShadow/DustMite#66] [Ideas]
There is already

```
  --coverage DIR     Load .lst files corresponding to source files from DIR
```

. Why doesn't `dub dustmite` generate new coverage information "at each call" to `dub run` and use that via `loadCoverage` when deciding which code to remove prior to subsequent call?!!! Wouldn't this enable amazingly faster reduction speed for D especially in the reduction of dependencies.

I guess most of this logic should be implemented in `dub`.

Ping, @CyberShadow.

See https://github.com/CyberShadow/DustMite/blob/master/dustmite.d#L2405

**Comment by CyberShadow** [comment]
What are you trying to achieve with `--coverage`?

Can you describe your use case in more detail?

**Comment by nordlow** [comment]
I thought `loadCoverage()` gave hints to the reduction algorithm. Maybe I was wrong. What's the meaning of `loadCoverage`? Anyway, wouldn't information about unused lines and functions would be of exceptional value to the reduction algorithm?; the algorithm could start with remove the single lines that aren't covered typically inside conditionals and then try removing functions that were called in covered lines... Are you following?

**Comment by CyberShadow** [comment]
`--coverage` makes DustMite avoid reducing code that ran. Generally that set of code shouldn't change. It's possible that it does change, e.g. if DustMite reduces some non-code declarations that cause some code to no longer run, but I can't think of a practical use case for this scenario.

**Comment by CyberShadow** [comment]
> the algorithm could start with remove the single lines that aren't covered typically inside conditionals and then try removing functions that were called in covered lines... Are you following?

In many cases it's just as useful to try reducing code that ran as everything else, so I can't really think of a good argument in favor of that.

**Reply by nordlow** [reply]
What about telling the reduction algorithm to instead try reducing code that didn't run or is that equivalent to the current behaviour?

**Reply by CyberShadow** [reply]
Yes, that's what `--coverage` does. It prevents code that ran from being reduced. Usually it's not what you want, though.

**Reply by nordlow** [reply]
Ok, let me revise. The reduction algorithm should _prioritize_ reducing uncovered lines _before_ potentially reducing covered ones. I believe that behavior should be the default for D code having coverage data extracted from recent run. Afaict, this is gonna lead to an fantastic speed up of the reduction especially when when reducing dependencies using `dub dustmite`. The important thing is to first (as soon as possible) remove uncovered code and secondary try reducing covered code. How do I modify the reduction algorithm to make use of the coverage data in this alternative way?

**Reply by CyberShadow** [reply]
You can do that by running DustMite with `--coverage`, and then again without it.

**Reply by nordlow** [reply]
Ok, thanks. The administration of producing and reading the .lst-files needs to be added to `dub dustmite`, though.

**Reply by CyberShadow** [reply]
I don't even remember why it was added.

I'm not going to make any changes without some hard evidence that they're beneficial :)

**Reply by nordlow** [reply]
> I don't even remember why it was added.
> 
> I'm not going to make any changes without some hard evidence that they're beneficial :)

What kinds of DustMite-changes are you thinking of; automatic generation and reading of .lst files when `--coverage` without arguments is passed?

**Reply by CyberShadow** [reply]
Oh, I misunderstood what you said. I thought you said that there was something that needed to be added to DustMite itself.

**Comment by nordlow** [comment]
Btw, my dustmite reduction takes too long time reducing packages. What strategy should I try then?

**Comment by CyberShadow** [comment]
Have you tried `--strategy=careful` yet?

**Comment by nordlow** [comment]
Now. I'll try that. Can I abort a DustMite with Ctrl-C and then use the the result as input to a new dustmite?

**Comment by CyberShadow** [comment]
Of course (assuming you didn't use `--no-save`).

**Comment by nordlow** [comment]
I first needed https://github.com/dlang/dub/pull/2173.

**Comment by nordlow** [comment]
Btw, does dustmite utilize all the cores when checking if a set of tried splits succeeds in create a program the compiles? And does `dub dustmite` do the same?

**Comment by nordlow** [comment]
Btw, at https://dlang.org/blog/2020/04/13/dustmite-the-general-purpose-data-reduction-tool/ I read

"The parser DustMite uses for D source code is intentionally very simple because it needs to handle potentially invalid D code, and you don’t want your bug reduction tool to also crash on top of the compiler."

? If the compilation always passes after every reduction why does parser need to handle invalid D code? And what is meant by invalid code in this case

**Comment by nordlow** [comment]
Btw, is there a node deletion history written to disk during the call to DustMite?

**Comment by CyberShadow** [comment]
> Btw, does dustmite utilize all the cores when checking if a set of tried splits succeeds in create a program the compiles?

I didn't understand that beyond "all cores".

You can ask DustMite to use all cores by using the `-j` flag.

No idea about the Dub side.

**Reply by nordlow** [reply]
> No idea about the Dub side.

Can I figure that out by looking at the output of `--trace`?

**Reply by CyberShadow** [reply]
I don't understand what you're trying to achieve here, can you reword?

As for `-j`, you would have to look if `dub` allows passing that flag to DustMite somehow.

**Comment by CyberShadow** [comment]
> ? If the compilation always passes after every reduction why does parser need to handle invalid D code?

It doesn't know that. You might be reducing a bug where compilation fails due to the bug you're trying to create a test case for.

**Comment by CyberShadow** [comment]
> Btw, is there a node deletion history written to disk during the call to DustMite?

There is `--trace` which will save verbose snapshots of every reduction attempt and whether it succeeded.

**Reply by nordlow** [reply]
Isn't it better to have a short description of what kind of reduction it tried?

**Reply by CyberShadow** [reply]
Why? What are you hoping to achieve by this?

**Reply by nordlow** [reply]
Say that you potentially have infinite set of possible reductions to try out of which only a small part compiles. If so and there are N cores at our disposal we should try syntax-checking (compiling) N of these pendings reductions in parallel and filter out the ones that compiles. If there's only a very small subset of these reductions that pass and D's parser is very fast this is gonna result in an enormous speedup. I assume this is already implemented, right? If so what parameters describe a reduction; deleted region `R` (start end end byte offset) in file `F`?

**Reply by CyberShadow** [reply]
> I assume this is already implemented, right?

Yes, if you use `-j`.

> If so what parameters describe a reduction; deleted region `R` (start end end byte offset) in file `F`?

The `Reduction` struct describes a reduction:
https://github.com/CyberShadow/DustMite/blob/6634109faa2c53ab8d1f0acd8c06f0793ad1ec4d/dustmite.d#L58

but I don't understand how that relates to the rest of your post.

**Comment by nordlow** [comment]
Ahh, I misinterpreted

  -j[N]              Use N look-ahead processes (48 by default)

as using 48 cores when even I don't specify `-j`. Thanks for clarifying this.

**Comment by nordlow** [comment]
According to my measurements my `dub dustmite` reduction is currently spending most of its time compiling successful programs where only a very small fraction of the number of lines are removed at every step. Making use of the last profiling information to guide the next reduction to remove lines not visited would afaict result in a much greater relative reduction of number of lines each time. I still don't understand why this isn't considered.

Would it be possible for you to experiment with running `dub dustmite` using https://github.com/dlang/dub/pull/2174?

I'm currently using

```sh
dub dustmite -c it -b unittest --compiler=ldc2 --program-status=1 --program-regex='stack-buffer-overflow' --no-redirect -j32 --strategy=careful OUTPUT_DIR -- -s -d
```

as the error is triggered in unit-threaded unittest execution.

**Reply by CyberShadow** [reply]
I still don't understand what you want DustMite to do, and why you think that will work. Can you make an example?

Are you using `--strategy=careful` yet?

**Reply by nordlow** [reply]
> I still don't understand what you want DustMite to do, and why you think that will work. Can you make an example?
> 
> Are you using `--strategy=careful` yet?

I am.

**Reply by nordlow** [reply]
Note that I'm working via `dub dustmite` using dub master. I'm working on adding support for forwarding of dustmite flags. See: 

- https://github.com/dlang/dub/pull/2174
- https://github.com/dlang/dub/pulls?q=is%3Apr+is%3Amerged

**Reply by nordlow** [reply]
I have checked that the -jN is correctly forwarded from dub dustmite to dustmite on my branch at dlang/dub#2174 but for some reason I only get one concurrent compiler process running at the same time in my process overview. I don't understand why. Together with `--strategy=careful` that is.

**Reply by CyberShadow** [reply]
Are you seeing `(lookahead)` or `(lookahead-wait)` in the log?

**Reply by nordlow** [reply]
Where is the log?

**Reply by CyberShadow** [reply]
In your terminal.

**Reply by nordlow** [reply]
No, I'm not seeing any instance of `lookahead`. In the output. Have checked both stdout and stderr.

**Reply by CyberShadow** [reply]
Then `-j` is probably not used. Check the DustMite command line with a process viewer.

Also make sure you are using a recent version of DustMite from this repository.

**Reply by nordlow** [reply]
Ahh, some instance get the `-j32` flag some don't because the dub in the path isn't the `-j` enabled.

When dub calls itself does it call dub in the path or the original dub executable? I'm guessing the one in the PATH which isn't -j enabled. I presume it should call itself using the original binary, right?

**Reply by nordlow** [reply]
I had some old dustmite processes left. I've checked now that all calls to dustmite receive the -j32 flag. So I still don't know what's wrong. Still only one simulateneous call to ldc2 from dustmite. And still no output matching `lookahead`. I'm using dustmite master. Could it be because I have a lot dub deps?

**Reply by CyberShadow** [reply]
Your terminal output should look something like this:

```
None => Yes
############### ITERATION 0 ################
============= Depth 0 =============
[ 0.0%] Remove [] => No (lookahead-wait)
============= Depth 1 =============
[ 0.0%] Remove [0] => No (cached)
[ 0.0%] Concat [0] => N/A
============= Depth 2 =============
[ 0.0%] Remove [00] => No (lookahead-wait)
============= Depth 3 =============
[ 0.0%] Remove [001] => No (lookahead-wait)
[ 2.5%] Remove [000] => No (cached)
============= Depth 4 =============
[ 0.0%] Remove [0010] => No (cached)
[ 2.5%] Remove [0001] => No (lookahead)
[93.6%] Remove [0000] => No (lookahead)
[93.6%] Unwrap [0000] => No (lookahead)
============= Depth 5 =============
[ 0.0%] Remove [00101] => Yes (lookahead-wait)
[ 6.6%] Remove [00100] => No (cached)
[ 8.6%] Remove [00011] => No (lookahead)
[78.9%] Remove [00010] => No (lookahead)
[93.6%] Remove [00000] => No (lookahead)
############### ITERATION 1 ################
============= Depth 0 =============
[ 0.0%] Remove [] => No (cached)
============= Depth 1 =============
[ 0.0%] Remove [0] => No (cached)
[ 0.0%] Concat [0] => N/A
============= Depth 2 =============
[ 0.0%] Remove [00] => No (cached)
============= Depth 3 =============
[ 0.0%] Remove [001] => No (cached)
[ 2.1%] Remove [000] => No (cached)
============= Depth 4 =============
[ 0.0%] Remove [0010] => No (cached)
[ 2.1%] Remove [0001] => No (cached)
[93.1%] Remove [0000] => No (lookahead)
[93.1%] Unwrap [0000] => No (lookahead)
============= Depth 5 =============
[ 0.0%] Remove [00100] => No (cached)
[ 2.1%] Remove [00011] => No (cached)
[77.4%] Remove [00010] => No (cached)
[93.1%] Remove [00000] => No (cached)
============= Depth 6 =============
[ 0.0%] Remove [001001] => Yes (lookahead)
[ 4.1%] Remove [001000] => No (cached)
[ 5.3%] Remove [000110] => No (lookahead-wait)
[79.2%] Remove [000101] => No (lookahead-wait)
[79.3%] Remove [000100] => No (lookahead)
[93.1%] Remove [000001] => No (lookahead)
[98.0%] Remove [000000] => No (lookahead)
############### ITERATION 2 ################
...
```

**Reply by nordlow** [reply]
I'm gonna try setting up a open repo that replicates the problem...

**Reply by nordlow** [reply]
Btw, have you actually ever studied the behavior and output of dub dustmite? I have the feeling it's doing something severly incorrect.

**Reply by nordlow** [reply]
I get 

```
None => Yes
############### ITERATION 0 ################
============= Depth 0 =============
[ 0.0%] Remove [] => No
============= Depth 1 =============
[ 0.0%] Remove [1] => No
[15.3%] Remove [0] => No
============= Depth 2 =============
[ 0.0%] Remove [11] => No
[ 8.1%] Remove [10] => Yes
[15.3%] Remove [01] => No
[38.2%] Remove [00] => No
============= Depth 3 =============
[ 0.0%] Remove [101] => Yes
[ 3.3%] Remove [100] => No (cached)
[ 8.8%] Remove [011] => No
[20.3%] Remove [010] => No
[33.5%] Remove [001] => No
[63.4%] Remove [000] => No
============= Depth 4 =============
[ 0.0%] Remove [1001] => No
[ 3.9%] Remove [1000] => Yes
[ 5.6%] Remove [0111] => Yes
[10.4%] Remove [0110] => No
[17.6%] Remove [0101] => No
[24.4%] Remove [0100] => No
[31.1%] Remove [0011] => No
[39.8%] Remove [0010] => No
[62.1%] Remove [0001] => No
[93.4%] Remove [0000] => No
============= Depth 5 =============
[ 0.0%] Remove [10001] => No
[ 2.4%] Remove [10000] => No
[ 4.2%] Remove [01101] => No
[ 9.8%] Remove [01100] => No
[11.8%] Remove [01011] => No
[15.5%] Remove [01010] => No
[19.1%] Remove [01001] => No
[26.0%] Remove [01000] => No
[26.4%] Remove [00111] => No
[29.6%] Remove [00110] => No
[35.7%] Remove [00101] => No
[46.8%] Remove [00100] => No
[59.5%] Remove [00011] => No
[90.3%] Remove [00010] => No
[92.9%] Remove [00001] => No
[96.8%] Remove [00000] => No
============= Depth 6 =============
[ 0.0%] Remove [100011] => No
[ 0.8%] Remove [100010] => No
[ 2.4%] Remove [100001] => No
[ 2.9%] Remove [100000] => Yes
[ 4.2%] Remove [011011] => No
[ 4.4%] Remove [011010] => No
[ 9.8%] Remove [011001] => Yes
[10.5%] Remove [011000] => No
[11.8%] Remove [010111] => No
[13.5%] Remove [010110] => No
[15.5%] Remove [010101] => No
[18.3%] Remove [010100] => No
[19.1%] Remove [010011] => No
[22.8%] Remove [010010] => No
[26.0%] Remove [010001] => No
[26.4%] Remove [010000] => Yes
[26.4%] Remove [001111] => Yes
[26.4%] Remove [001110] => No
[29.6%] Remove [001101] => No
[35.5%] Remove [001100] => 
```

**Reply by CyberShadow** [reply]
> Btw, have you actually ever studied the behavior and output of dub dustmite? I have the feeling it's doing something severly incorrect.

No. I didn't write it and never looked at it. If someone has an issue with Dub and DustMite, generally I tell them to first completely remove Dub from the equation.

**Reply by CyberShadow** [reply]
> I get

Looks to me like your `-j` is not working.

**Reply by nordlow** [reply]
Here's a snapshot of the contents of M-x proced on Emacs:

```
per   310427   2       0.0  0.0  16:47    00:00 /home/per/Work/dub/bin/dub dustmite -c it -b unittest --compiler=ldc2 --program-status=1 --program-regex=stack-buffer-overflow -j32 --strategy=careful ../PACKAGE_DIR.reduce.1 -- -s -d
per   310432    3      2.1  1.5  16:47    02:42 /home/per/.local/dlang/linux/bin64/dustmite /home/per/Work/PACKAGE_DIR.reduce.1 /home/per/Work/dub/bin/dub\ dustmite\ --test-package=PACKAGE_NAME\ --build=unittest\ --config=it\ "--compiler=ldc2"\ --program-status=1\ "--program-regex=stack-buffer-overflow"\ "--strategy=careful"\ "-j32"\ --vquiet
per  1133594     4     0.2  0.0  18:53    00:00 /bin/sh -c /home/per/Work/dub/bin/dub\ dustmite\ --test-package=PACKAGE_NAME\ --build=unittest\ --config=it\ "--compiler=ldc2"\ --program-status=1\ "--program-regex=stack-buffer-overflow"\ "--strategy=careful"\ "-j32"\ --vquiet
per  1133595      5    0.4  0.0  18:53    00:00 /home/per/Work/dub/bin/dub dustmite --test-package=PACKAGE_NAME --build=unittest --config=it --compiler=ldc2 --program-status=1 --program-regex=stack-buffer-overflow --strategy=careful -j32 --vquiet
```

. Do you see anything faulty?

The `/home/per/Work/dub/bin/dub` is the `-j` enabled dub.

**Reply by CyberShadow** [reply]
Yes. The second command has `-j` in the test command (dub passing `-j` back to itself), but not as a parameter to DustMite itself.

**Reply by CyberShadow** [reply]
BTW, the same is true with `--strategy=careful`, and I can see from the output you posted above that it's still using the default strategy.

**Reply by nordlow** [reply]
Thanks!

**Reply by nordlow** [reply]
This solves my problems https://github.com/dlang/dub/pull/2175.

Output now looks more like ours:

```
Yes
############### ITERATION 0 ################
============= Depth 0 =============
[ 0.0%] Remove [] => No (lookahead-wait)
============= Depth 1 =============
[ 0.0%] Remove [1] => No (lookahead-wait)
[15.3%] Remove [0] => No (lookahead)
============= Depth 2 =============
[ 0.0%] Remove [11] => No (lookahead)
[ 8.1%] Remove [10] => Yes (lookahead-wait)
[15.3%] Remove [01] => No (lookahead)
[38.2%] Remove [00] => No (lookahead)
############### ITERATION 1 ################
============= Depth 0 =============
[ 0.0%] Remove [] => No (cached)
============= Depth 1 =============
[ 0.0%] Remove [1] => No (cached)
[ 8.8%] Remove [0] => No (lookahead)
============= Depth 2 =============
[ 0.0%] Remove [10] => No (cached)
[ 8.8%] Remove [01] => No (cached)
[33.5%] Remove [00] => No (cached)
============= Depth 3 =============
[ 0.0%] Remove [101] => Yes (lookahead)
[ 3.3%] Remove [100] => No (cached)
[ 8.8%] Remove [011] => No (lookahead)
[20.3%] Remove [010] => No (lookahead)
[33.5%] Remove [001] => No (lookahead)
[63.4%] Remove [000] => No (lookahead)
############### ITERATION 2 ################
============= Depth 0 =============
[ 0.0%] Remove [] => No (cached)
============= Depth 1 =============
[ 0.0%] Remove [1] => No (cached)
[ 5.6%] Remove [0] => No (lookahead)
============= Depth 2 =============
[ 0.0%] Remove [10] => No (cached)
[ 5.6%] Remove [01] => No (lookahead)
[31.1%] Remove [00] => No (lookahead)
============= Depth 3 =============
[ 0.0%] Remove [100] => No (cached)
[ 5.6%] Remove [011] => No (cached)
[17.6%] Remove [010] => No (cached)
[31.1%] Remove [001] => No (cached)
[62.1%] Remove [000] => No (cached)
============= Depth 4 =============
[ 0.0%] Remove [1001] => No (lookahead)
[ 3.9%] Remove [1000] => Yes (lookahead-wait)
[ 5.6%] Remove [0111] => Yes (lookahead-wait)
[10.4%] Remove [0110] => No (lookahead)
[17.6%] Remove [0101] => No (lookahead)
[24.4%] Remove [0100] => No (lookahead)
[31.1%] Remove [0011] => No (lookahead)
[39.8%] Remove [0010] => No (lookahead)
[62.1%] Remove [0001] => No (lookahead)
```

**Comment by nordlow** [comment]
Btw, have you seen [creduce](https://github.com/csmith-project/creduce) having activity as recent as 21 of June 2021.

**Comment by nordlow** [comment]
Moreover I believe the docs should provide a link to https://en.wikipedia.org/wiki/Delta_debugging and possibly also https://news.ycombinator.com/item?id=22855633.
